File: Problems/230_Kth_Smallest_Element_in_a_BST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# c;ean solution
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        curr = root
        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            k -= 1
            if k == 0:
                return curr.val
            curr = curr.right

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        stack = []
        curr = root
        stack.append(curr)
        
        while stack:
            while curr:
                stack.append(curr)
                curr = curr.left

            curr = stack.pop()
            k -= 1
            if k == 0:
                return curr.val
            curr = curr.right
            
# class Solution:
#     def __init__(self):
#         self.k = None
#         self.ans = None

#     def kthSmallest(self, root: TreeNode, k: int) -> int:
#         self.k = k
#         self.traverse(root)
#         return self.ans

#     def traverse(self, node):
#         if not node or self.k<0:
#             return
#         self.traverse(node.left)
#         self.k -= 1
#         if self.k == 0:
#             self.ans = node.val
#             return 
#         self.traverse(node.right)

File: Problems/230_Kth_Smallest_Element_in_a_BST.cpp
#include <stack>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int &k) {
        if(!root) return -1;
        int res = kthSmallest(root->left, k);
        if(res!=-1) return res;
        k--;
        if(k==0) return root->val;
        return kthSmallest(root->right, k); 
    }
};

class Solution {
public:
    int K;
    int kthSmallest(TreeNode* root, int k) {
        K = k;
        return traverse(root);
    }
    int traverse(TreeNode* node){
        if(!node) return -1;
        int res = traverse(node->left);
        if(res!=-1) return res;
        K--;
        if(K==0) return node->val;
        return traverse(node->right);
    }
};

class Solution {
public:
    int count;
    int kthSmallest(TreeNode* root, int k){
        stack<TreeNode*> st;
        st.push(root);
        TreeNode* curr = root;
        while(!st.empty() || curr){
            while(curr){
                st.push(curr);
                curr = curr->left;
            }
            curr = st.top();
            st.pop();
            k--;
            if(k==0){
                return curr->val;
            }
            curr = curr->right;
        }
        return -1;
    }
};